plugins/chat/assets/javascripts/discourse/models/chat-message.js

```diff
@@ -1,26 +1,245 @@
+      Math.abs(new Date(this.createdAt) - new Date(this.createdAt)) < 300000 && // If the time between messages is over 5 minutes, break.
+      this.user.id === this.previousMessage?.user?.id
+    );
+  }
+
+  get firstMessageOfTheDayAt() {
+    if (!this.previousMessage) {
       return;
     }
 
-    args.user = User.create(args.user);
-  },
-});
+    if (
+      !this.#areDatesOnSameDay(
+        new Date(this.previousMessage.createdAt),
+        new Date(this.createdAt)
+      )
+    ) {
+      return moment(this.createdAt).calendar(moment(), {
+        sameDay: `[${I18n.t("chat.chat_message_separator.today")}]`,
+        lastDay: `[${I18n.t("chat.chat_message_separator.yesterday")}]`,
+        lastWeek: "LL",
+        sameElse: "LL",
+      });
+    }
+  }
+
+  @cached
+  get index() {
+    return this.channel.messages.indexOf(this);
+  }
+
+  @cached
+  get previousMessage() {
+    return this.channel?.messages?.objectAt?.(this.index - 1);
+  }
+
+  @cached
+  get nextMessage() {
+    return this.channel?.messages?.objectAt?.(this.index + 1);
+  }
+
+  get staged() {
+    return this.stagedId?.length > 0;
+  }
+
+  react(emoji, action, actor, currentUserId) {
+    const selfReaction = actor.id === currentUserId;
+    const existingReaction = this.reactions.find(
+      (reaction) => reaction.emoji === emoji
+    );
+
+    if (existingReaction) {
+      if (action === "add") {
+        if (selfReaction && existingReaction.reacted) {
+          return false;
+        }
+
+        existingReaction.count = existingReaction.count + 1;
+        if (selfReaction) {
+          existingReaction.reacted = true;
+        }
+        existingReaction.users.pushObject(actor);
+      } else {
+        existingReaction.count = existingReaction.count - 1;
+
+        if (selfReaction) {
+          existingReaction.reacted = false;
+        }
+
+        if (existingReaction.count === 0) {
+          this.reactions.removeObject(existingReaction);
+        } else {
+          existingReaction.users.removeObject(
+            existingReaction.users.find((user) => user.id === actor.id)
+          );
+        }
+      }
+    } else {
+      if (action === "add") {
+        this.reactions.pushObject(
+          ChatMessageReaction.create({
+            count: 1,
+            emoji,
+            reacted: selfReaction,
+            users: [actor],
+          })
+        );
+      }
+    }
+  }
+
+  #initChatMessageReactionModel(messageId, reactions = []) {
+    return reactions.map((reaction) =>
+      ChatMessageReaction.create(Object.assign({ messageId }, reaction))
+    );
+  }
+
+  #initUserModel(user) {
+    if (!user || user instanceof User) {
+      return user;
+    }
+
+    return User.create(user);
+  }
+
+  #areDatesOnSameDay(a, b) {
+    return (
+      a.getFullYear() === b.getFullYear() &&
+      a.getMonth() === b.getMonth() &&
+      a.getDate() === b.getDate()
+    );
+  }
+}
+
+// https://stackoverflow.com/a/8809472
+function generateUDID() {
```

Comment:
We already have this in the codebase https://github.com/discourse/discourse/blob/61e2131738fde0dc72dd03d5b091d94dbf559e72/app/assets/javascripts/pretty-text/addon/guid.js . Since these pretty-text helpers can import from discourse-common/lib, we should probably make a new file for this to export the `generateUUID()` function, similar to https://github.com/discourse/discourse/blob/61e2131738fde0dc72dd03d5b091d94dbf559e72/app/assets/javascripts/discourse-common/addon/lib/dom-from-string.js

Reply:
Hmm nice didn't know, will look to use it :+1: